### DISCUSSION TITLE
Multiple Classes with Semantic Segmentation

### DISCUSSION THREAD

**Discussion by natedailey7** [azavea/raster-vision#2079] [Q&A] [answered]
I'm trying to input multiple classes into my semantic segmentation model, and I'm having trouble passing my class information into my `SemanticSegmentationSlidingWindowGeoDataset`. Previously, I asked [the same question](https://github.com/azavea/raster-vision/discussions/2027) about `ObjectDetectionSlidingWindowGeoDataset`'s, but I'm having trouble migrating this technique to semantic segmentation. 

Here's what I'm trying:
```python
class_config = ClassConfig(
    names=['class1', 'class2', 'class3'],
    colors=['lightgreen', 'lightblue', 'darkred'],
    )

train_image_uri = "path"
train_label_uri = "path"
val_image_uri   = "path"
val_label_uri   = "path"

class SemanticSegmentationSlidingWindowGeoDatasetCustom(SemanticSegmentationSlidingWindowGeoDataset):
	def __getitem__(self, key):
		x, y = super().__getitem__(key)
		y = y.long()
		return x, y

train_ds = SemanticSegmentationSlidingWindowGeoDatasetCustom.from_uris(
    class_config=class_config,
    image_uri=train_image_uri,
    label_vector_uri=train_label_uri,
    label_vector_source_kw=dict(
          transformers=[ClassInferenceTransformerConfig(
              class_id_to_filter={0: ['==', 'id', 0], 1: ['==', 'id', 1], 3: ['==', 'id', 3]})]),  # where 'id' is the name of the field containing my classes in the .geojson
    size=400,
    stride=400,
    transform=A.Resize(400, 400)
)
```
Which yields the error:
```
TypeError: GeoJSONVectorSource.__init__() got an unexpected keyword argument 'transformers'
```

After reading the documentation for `SemanticSegmentationSlidingWindowGeoDataset` and `GeoJSONVectorSource`, I'm still not sure where to place my `ClassInferenceTransformerConfig`. Any tips would be greatly appreciated.

**Comment by AdeelH** [comment, marked as answer]
Yeah, I agree that is confusing. This is because of a discrepancy between the implementations of [`make_od_scene`](https://docs.rastervision.io/en/stable/api_reference/_generated/rastervision.core.data.utils.factory.make_od_scene.html#rastervision.core.data.utils.factory.make_od_scene) and [`make_ss_scene`](https://docs.rastervision.io/en/stable/api_reference/_generated/rastervision.core.data.utils.factory.make_ss_scene.html). For semantic segmentation, you'll have to do it like this:

```py
train_ds = SemanticSegmentationSlidingWindowGeoDatasetCustom.from_uris(
    ...
    label_vector_source_kw=dict(
          vector_transformers=[ClassInferenceTransformer(
              default_class_id=None,
              class_id_to_filter={0: ['==', 'id', 0], 1: ['==', 'id', 1], 3: ['==', 'id', 3]})]), 
    ...
)
```

- `transformers` --> `vector_transformers`
- `ClassInferenceTransformerConfig` --> `ClassInferenceTransformer`

**Reply by natedailey7** [reply]
This worked perfectly for me, thanks so much!
